Scheduler.lib.nut

```diff
@@ -0,0 +1,430 @@
+//MIT License
+//
+//Copyright 2017 Mystic Pants
+//
+//SPDX-License-Identifier: MIT
+//
+//Permission is hereby granted, free of charge, to any person obtaining a copy
+//of this software and associated documentation files (the "Software"), to deal
+//in the Software without restriction, including without limitation the rights
+//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//copies of the Software, and to permit persons to whom the Software is
+//furnished to do so, subject to the following conditions:
+//
+//The above copyright notice and this permission notice shall be
+//included in all copies or substantial portions of the Software.
+//
+//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+//EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+//MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+//EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+//OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+//ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+//OTHER DEALINGS IN THE SOFTWARE.
+
+const JOB_TYPE_SET         = "set";
+const JOB_TYPE_AT          = "at";
+const JOB_TYPE_REPEAT      = "repeat";
+const JOB_TYPE_REPEAT_FROM = "repeat from";
+
+const JOB_ERROR_RESET = "Cannot reset job type: %s";
+
+class Scheduler {
+
+    static VERSION = "1.0.0";
+
+    _env = null;
+
+    _jobs       = [];
+    _currentJob = null;
+    _currentId  = null;
+    _nextId     = 1;
+
+    constructor() {
+        _env = imp.environment();
+    }
+
+    // Start a new timer to trigger after a specific duration
+    //
+    // Parameters:
+    //     dur (float)       the duration of the timer
+    //     cb  (function)    the function to run when the timer fires
+    // Return: (integer) the id number of the timer (can be used to cancel the timer)
+    function set(dur, cb, ...) {
+        local now = _getTime();
+        vargv.insert(0, null);
+
+        if (dur < 0) dur = 0;
+
+        local newJob = Job(this, {
+            "type": JOB_TYPE_SET,
+            "id": _nextId++,
+            "dur": dur,
+            "sec": _getSec(dur, now),
```

Comment:
I didn't quite get, why do we need both `sec` and `subSec`?

Reply:
So that the user can set floating point timers, e.g. 3.5 seconds.
They have to be in separate variables otherwise it will overflow.

Reply:
makes sense